patches/api/0446-Add-FluidState-API.patch

```diff
@@ -0,0 +1,201 @@
++    default boolean is(Fluid fluid) {
++        return getFluidType() == fluid;
++    }
++
++    /**
++     * Returns whether this fluid is a source block
++     *
++     * @return true if the fluid is a source block, false otherwise
++     */
++    boolean isSource();
++}
+diff --git a/src/main/java/io/papermc/paper/block/fluid/type/FallingFluid.java b/src/main/java/io/papermc/paper/block/fluid/type/FallingFluid.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..360a27fca76b0e63aa458aa3425a987224a4ac79
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/block/fluid/type/FallingFluid.java
+@@ -0,0 +1,13 @@
++package io.papermc.paper.block.fluid.type;
++
++import io.papermc.paper.block.fluid.FluidData;
++
++public interface FallingFluid extends FluidData {
++
++    /**
++     * Get if this liquid is falling.
++     *
++     * @return true if falling
++     */
++    boolean isFalling();
++}
+diff --git a/src/main/java/io/papermc/paper/block/fluid/type/Flowing.java b/src/main/java/io/papermc/paper/block/fluid/type/Flowing.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..14589b2de95e0f679fe9785e06648ace56449885
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/block/fluid/type/Flowing.java
+@@ -0,0 +1,26 @@
++package io.papermc.paper.block.fluid.type;
++
++public interface Flowing extends FallingFluid {
++
++    /**
++     * Get the level of the flowing liquid.
++     *
++     * @return the level
++     */
++    int getLevel();
++
++    /**
++     * Get the minimum possible level for the liquid.
++     *
++     * @return the minimum level
++     */
++    int getMinimumLevel();
++
++    /**
++     * Get the maximum possible level for the liquid.
++     *
++     * @return the maximum level
++     */
++    int getMaximumLevel();
++
++}
+diff --git a/src/main/java/org/bukkit/Fluid.java b/src/main/java/org/bukkit/Fluid.java
+index 7202b44a65e8699cf64329e441e83d4dd1779c19..8879f905d1c806ee35ac89ceebd346eea93913a7 100644
+--- a/src/main/java/org/bukkit/Fluid.java
++++ b/src/main/java/org/bukkit/Fluid.java
+@@ -17,19 +17,19 @@ public enum Fluid implements Keyed {
+     /**
+      * Stationary water.
+      */
+-    WATER,
++    WATER{public boolean isWater() {return true;}}, // Paper
+     /**
+      * Flowing water.
+      */
+-    FLOWING_WATER,
++    FLOWING_WATER{public boolean isWater() {return true;}}, // Paper
+     /**
+      * Stationary lava.
+      */
+-    LAVA,
++    LAVA{public boolean isLava() {return true;}}, // Paper
+     /**
+      * Flowing lava.
+      */
+-    FLOWING_LAVA;
++    FLOWING_LAVA{public boolean isLava() {return true;}}; // Paper
+ 
+     private final NamespacedKey key;
+ 
+@@ -42,4 +42,13 @@ public enum Fluid implements Keyed {
+     public NamespacedKey getKey() {
+         return key;
+     }
++    // Paper start
++    public boolean isWater() {
++        return false;
++    }
++
++    public boolean isLava() {
++        return false;
++    }
++    // Paper end
```

Comment:
These are incorrect. I don't think *any* isWhatever methods should be added. These are already set by tags and changeable by datapacks. Those tags are already exposed to the API, I don't think we need an alternative way to do it.

Reply:
Ok! Will be removed